### PR TITLE
fix: memory leak in unsafe code

### LINKED ISSUE #12445: bug: memory leak in unsafe code
**Summary**
https://github.com/datafuselabs/databend/blob/b6cb56baa276b9c0c794fd83ec3851eeba5c0252/src/query/expression/src/block.rs#L90-L114
The memory is not freed after `Box::into_raw()`

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #12445

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12446)

### CODE REVIEW

**Comment by dantengsky** [comment]
`jeprof` find no memory with this PR, for the simulated copy-into test scenario.


2023-08-12-183340_1115x91_scrot

**Review by dantengsky — APPROVED** [review]
let's merge

File: src/query/expression/src/block.rs [hunk 1]
```diff
@@ -92,24 +92,16 @@ impl<T: BlockMetaInfo> BlockMetaInfoDowncast for T {
         if boxed.as_any().is::<T>() {
             unsafe {
                 // SAFETY: `is` ensures this type cast is correct
-                let raw_ptr = Box::into_raw(boxed) as *const dyn BlockMetaInfo;
-                return Some(std::ptr::read(raw_ptr as *const Self));
+                let raw_ptr = Box::into_raw(boxed) as *mut dyn BlockMetaInfo;
+                let typed_ptr = raw_ptr as *mut Self;
+                return Some(*Box::from_raw(typed_ptr));
```
Comment (andylokandy) [inline src/query/expression/src/block.rs:97]: it'll be more consice if we write:
```
let downcast_box = std::mem::transmute::<Box<_>, Box< Self>>(boxed);
*downcast_box
```

Comment (SkyFan2002) [reply src/query/expression/src/block.rs:97]: Good idea! BTW, I want to discuss, what about removing unsafe code by writing somthing like `boxed.into_any().downcast()`

Comment (andylokandy) [reply src/query/expression/src/block.rs:97]: it may be not able to take ownership from Box<Any>, but we can clone it anyway.

**Pull request merged**